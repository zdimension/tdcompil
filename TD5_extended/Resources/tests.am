# Optimizations enabled
NEW "generated" 4		# line 1536
START @INIT		# line 1537
END @END "END"		# line 1538
END @DEFAULT "default option"		# line 1539
# Memory map
# POSITION  SIZE  NAME
#        0     1  y
FROM @INIT		# line 1553
'_,'_,'_,'_ '[,'[,'_,'_ R,S,S,S @0		# line 1554
# allocate variable y (position 0, size 1)		# line 1558
FROM @0		# line 1575
'_,'[,'_,'_ '0,'[,'_,'_ R,S,S,S @1		# line 1576
FROM @1		# line 1575
'_,'[,'_,'_ '0,'[,'_,'_ R,S,S,S @2		# line 1576
FROM @2		# line 1575
'_,'[,'_,'_ '0,'[,'_,'_ R,S,S,S @3		# line 1576
FROM @3		# line 1575
'_,'[,'_,'_ '0,'[,'_,'_ R,S,S,S @4		# line 1576
FROM @4		# line 1575
'_,'[,'_,'_ '0,'[,'_,'_ R,S,S,S @5		# line 1576
FROM @5		# line 1575
'_,'[,'_,'_ '0,'[,'_,'_ R,S,S,S @6		# line 1576
FROM @6		# line 1575
'_,'[,'_,'_ '0,'[,'_,'_ R,S,S,S @7		# line 1576
FROM @7		# line 1575
'_,'[,'_,'_ '0,'[,'_,'_ R,S,S,S @8		# line 1576
FROM @8		# line 1579
'_,'[,'_,'_ '/,'[,'_,'_ R,S,S,S @9		# line 1580
FROM @9		# line 1583
'_|'/|'0|'1,'[,'_,'_ L,S,S,S		# line 1584
'[,'[,'_,'_ S,S,S,S @10		# line 1585
#	------ BEGIN: program
# KIND = 2 (k_operator)		# line 428
#	; first
# KIND = 2 (k_operator)		# line 428
#	store
#	evaluating right-hand value
# KIND = 1 (k_number)		# line 428
#	statically evaluated
#	push	5
FROM @10		# line 155
'/|'[,'[,'_,'_ S,R,S,S @11		# line 156
'/|'[,'/,'_,'_ S,R,S,S @11		# line 157
FROM @11		# line 158
'/|'[,'_,'_,'_ '/|'[,'1,'_,'_ S,R,S,S @12		# line 161
FROM @12		# line 162
'/|'[,'_,'_,'_ '/|'[,'0,'_,'_ S,R,S,S @13		# line 161
FROM @13		# line 162
'/|'[,'_,'_,'_ '/|'[,'1,'_,'_ S,R,S,S @14		# line 161
FROM @14		# line 162
'/|'[,'_,'_,'_ '/|'[,'0,'_,'_ S,R,S,S @15		# line 161
FROM @15		# line 162
'/|'[,'_,'_,'_ '/|'[,'0,'_,'_ S,R,S,S @16		# line 161
FROM @16		# line 162
'/|'[,'_,'_,'_ '/|'[,'0,'_,'_ S,R,S,S @17		# line 161
FROM @17		# line 162
'/|'[,'_,'_,'_ '/|'[,'0,'_,'_ S,R,S,S @18		# line 161
FROM @18		# line 162
'/|'[,'_,'_,'_ '/|'[,'0,'_,'_ S,R,S,S @19		# line 161
FROM @19		# line 162
'/|'[,'_,'_,'_ '/|'[,'/,'_,'_ S,S,S,S @20		# line 164
FROM @20		# line 1181
'0|'1|'/,'/,'_,'_ L,S,S,S		# line 1182
'[,'/,'_,'_ S,S,S,S @21		# line 1183
# navigating to y		# line 1319
#	navigating to left of stack head
FROM @21		# line 1186
'/|'[,'/,'_,'_ S,L,S,S @22		# line 1187
FROM @22		# line 1188
'/,'0|'1,'_,'_ S,L,S,S		# line 1189
'[,'0|'1,'_,'_ S,L,S,S		# line 1190
'/,'/|'[,'_,'_ R,R,S,S @23		# line 1191
'[,'/|'[,'_,'_ R,R,S,S @23		# line 1192
FROM @23		# line 1195
'0|'1,'0,'_,'_ '0,'_,'_,'_ R,R,S,S		# line 1196
'0|'1,'1,'_,'_ '1,'_,'_,'_ R,R,S,S		# line 1197
'/,'/,'_,'_ '/,'_,'_,'_ L,L,S,S @24		# line 1198
FROM @24		# line 1199
'/|'0|'1|'[,'_,'_,'_ L,L,S,S		# line 1200
'/|'0|'1|'[,'/,'_,'_ S,S,S,S @25		# line 1201
'/|'0|'1|'[,'[,'_,'_ S,S,S,S @25		# line 1202
FROM @25		# line 1203
'/|'0|'1,'[,'_,'_ L,S,S,S		# line 1204
'/|'0|'1,'/,'_,'_ L,S,S,S		# line 1205
'[,'/|'[,'_,'_ S,S,S,S @26		# line 1206
#	; second
# KIND = 2 (k_operator)		# line 428
#	print
# KIND = 2 (k_operator)		# line 428
#	statically evaluated
#	load	y
# navigating to y		# line 1319
FROM @26		# line 465
'/|'[,'/,'_,'_ R,R,S,S @27		# line 466
'/|'[,'[,'_,'_ R,R,S,S @27		# line 467
FROM @27		# line 468
'0|'1,'_,'_,'_ '0|'1,'0|'1,'_,'_ R,R,S,S		# line 469
'/,'_,'_,'_ '/,'/,'_,'_ L,S,S,S @28		# line 470
FROM @28		# line 471
'/|'0|'1,'/,'_,'_ L,S,S,S		# line 472
'[,'/,'_,'_ S,S,S,S @29		# line 473
#	------ END: program
FROM @29		# line 1593
'[,'/,'_,'_ S,S,S,S @END		# line 1594
'[,'[,'_,'_ S,S,S,S @END		# line 1595
