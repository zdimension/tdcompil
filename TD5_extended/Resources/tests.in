{
	var x = 17;

	if (var y = x >> 1)
	{
		print y;
	}
}