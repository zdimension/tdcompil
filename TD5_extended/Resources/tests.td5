{
	type fibo_iter = struct { a: u16; b: u16; };
	type iterator = interface
	{
		func hasNext(iter: self*): bool;
		func next(iter: self*): u16;
	};

	impl fibo_iter
	{
		func init(iter: fibo_iter*): void
		{
			(*iter).a = 0;
			(*iter).b = 1;
		}
	}
	impl iterator for fibo_iter
    {
		func hasNext(iter: fibo_iter*): bool
		{
			return true;
		}

		func next(iter: fibo_iter*): u16
		{
			var a = (*iter).a;
			var b = (*iter).b;
			(*iter).a = b;
			(*iter).b = a + b;
			return a;
		}
	}

	func skip_one(iter: iterator*): void
	{
		iter.next();
	}

	var iter: fibo_iter;
	iter.init();

	for (var val in iter)
	{
		if (val > 10)
			break;
		print val;
	}
}