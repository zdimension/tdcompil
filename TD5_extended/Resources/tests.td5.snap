>{
#^ source.rust punctuation.section.scope.begin.rust
>	func isqrt(x: u16): u16
#^^^^^^ source.rust
#      ^^^^^ source.rust entity.name.function.rust.call
#           ^ source.rust
#            ^ source.rust
#             ^ source.rust
#              ^ source.rust
#               ^^^ source.rust storage.type.rust
#                  ^ source.rust
#                   ^ source.rust
#                    ^ source.rust
#                     ^^^ source.rust storage.type.rust
>	{
#^ source.rust
# ^ source.rust punctuation.section.scope.begin.rust
>		var rem: u16, root: u16;
#^^^^^^^^^ source.rust
#         ^ source.rust
#          ^ source.rust
#           ^^^ source.rust storage.type.rust
#              ^ source.rust
#               ^^^^^ source.rust
#                    ^ source.rust
#                     ^ source.rust
#                      ^^^ source.rust storage.type.rust
#                         ^ source.rust
>
>		for (var i = 0; i < 8; i++)
#^^ source.rust
#  ^^^ source.rust keyword.control.rust
#     ^^^^^^^^ source.rust
#             ^ source.rust
#              ^ source.rust
#               ^ source.rust constant.numeric.integer.decimal.rust
#                ^^^^ source.rust
#                    ^ source.rust keyword.operator.rust
#                     ^ source.rust
#                      ^ source.rust constant.numeric.integer.decimal.rust
#                       ^^^ source.rust
#                          ^ source.rust keyword.operator.rust
#                           ^ source.rust keyword.operator.rust
#                            ^^ source.rust
>		{
#^^ source.rust
#  ^ source.rust punctuation.section.scope.begin.rust
>			root = root << 1;
#^^^^^^^^ source.rust
#        ^ source.rust keyword.operator.rust
#         ^^^^^^ source.rust
#               ^ source.rust keyword.operator.rust
#                ^ source.rust keyword.operator.rust
#                 ^ source.rust
#                  ^ source.rust constant.numeric.integer.decimal.rust
#                   ^^ source.rust
>			rem = (rem << 2) + (x >> 14);
#^^^^^^^ source.rust
#       ^ source.rust keyword.operator.rust
#        ^^^^^^ source.rust
#              ^ source.rust keyword.operator.rust
#               ^ source.rust keyword.operator.rust
#                ^ source.rust
#                 ^ source.rust constant.numeric.integer.decimal.rust
#                  ^^ source.rust
#                    ^ source.rust keyword.operator.rust
#                     ^^^^ source.rust
#                         ^ source.rust keyword.operator.rust
#                          ^ source.rust keyword.operator.rust
#                           ^ source.rust
#                            ^^ source.rust constant.numeric.integer.decimal.rust
#                              ^^^ source.rust
>			x = x << 2;
#^^^^^ source.rust
#     ^ source.rust keyword.operator.rust
#      ^^^ source.rust
#         ^ source.rust keyword.operator.rust
#          ^ source.rust keyword.operator.rust
#           ^ source.rust
#            ^ source.rust constant.numeric.integer.decimal.rust
#             ^^ source.rust
>			root++;
#^^^^^^^ source.rust
#       ^ source.rust keyword.operator.rust
#        ^ source.rust keyword.operator.rust
#         ^^ source.rust
>			if (root <= rem)
#^^^ source.rust
#   ^^ source.rust keyword.control.rust
#     ^^^^^^^ source.rust
#            ^ source.rust keyword.operator.rust
#             ^ source.rust keyword.operator.rust
#              ^^^^^^ source.rust
>			{
#^^^ source.rust
#   ^ source.rust punctuation.section.scope.begin.rust
>				rem -= root;
#^^^^^^^^ source.rust
#        ^ source.rust keyword.operator.rust
#         ^ source.rust keyword.operator.rust
#          ^^^^^^^ source.rust
>				root++;
#^^^^^^^^ source.rust
#        ^ source.rust keyword.operator.rust
#         ^ source.rust keyword.operator.rust
#          ^^ source.rust
>			}
#^^^ source.rust
#   ^ source.rust punctuation.section.scope.end.rust
>			else
#^^^ source.rust
#   ^^^^ source.rust keyword.control.rust
>			{
#^^^ source.rust
#   ^ source.rust punctuation.section.scope.begin.rust
>				root--;
#^^^^^^^^ source.rust
#        ^ source.rust keyword.operator.rust
#         ^ source.rust keyword.operator.rust
#          ^^ source.rust
>			}
#^^^ source.rust
#   ^ source.rust punctuation.section.scope.end.rust
>		}
#^^ source.rust
#  ^ source.rust punctuation.section.scope.end.rust
>
>		return root >> 1;
#^^ source.rust
#  ^^^^^^ source.rust keyword.control.rust
#        ^^^^^^ source.rust
#              ^ source.rust keyword.operator.rust
#               ^ source.rust keyword.operator.rust
#                ^ source.rust
#                 ^ source.rust constant.numeric.integer.decimal.rust
#                  ^^ source.rust
>	}
#^ source.rust
# ^ source.rust punctuation.section.scope.end.rust
>
>
>	print isqrt(32768);
#^^^^^^^ source.rust
#       ^^^^^ source.rust entity.name.function.rust.call
#            ^ source.rust
#             ^^^^^ source.rust constant.numeric.integer.decimal.rust
#                  ^^^ source.rust
>}
#^ source.rust punctuation.section.scope.end.rust